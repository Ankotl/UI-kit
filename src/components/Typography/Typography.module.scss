@import '../../styles/typography';

/* variants */

.Typography_h1 {
  @include typography-h1;
}

.Typography_h2 {
  @include typography-h2;
}

.Typography_h3 {
  @include typography-h3;
}

.Typography_h4 {
  @include typography-h4;
}

.Typography_p1 {
  @include typography-p1;
}

.Typography_p2 {
  @include typography-p2;
}

.Typography_p3 {
  @include typography-p3;
}

.Typography_caption {
  @include typography-caption;
}