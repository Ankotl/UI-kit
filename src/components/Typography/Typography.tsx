import React from 'react'

import clsx from 'clsx'

import { useIsMobile } from 'hooks/useMediaQuery/useMediaQuery'

import s from './Typography.module.scss'

// TODO move to type, not enum
export enum EnumTypographyVariants {
  H1 = 'h1',
  H2 = 'h2',
  H3 = 'h3',
  H4 = 'h4',
  P1 = 'p1',
  P2 = 'p2',
  P3 = 'p3',
  CAPTION = 'caption',
}

export const getDefaultTagFromVariant = (
  variant: EnumTypographyVariants
): React.ElementType => {
  if (variant === EnumTypographyVariants.CAPTION) {
    return 'div'
  }

  if (
    variant === EnumTypographyVariants.P1 ||
    variant === EnumTypographyVariants.P2 ||
    variant === EnumTypographyVariants.P3
  ) {
    return 'p'
  }

  return variant
}

type TypographyOwnProps<E extends React.ElementType> = {
  variant?: EnumTypographyVariants
  mobileVariant?: EnumTypographyVariants
  as?: E
  className?: string
  text?: string | number
}

export type ITypographyProps<E extends React.ElementType> =
  TypographyOwnProps<E> &
    Omit<React.ComponentProps<E>, keyof TypographyOwnProps<E>>

const TypographyInner = <E extends React.ElementType = 'div'>(
  {
    as,
    className,
    children,
    text,
    mobileVariant,
    variant = EnumTypographyVariants.H1,
    ...otherProps
  }: ITypographyProps<E>,
  ref: React.ForwardedRef<E>
) => {
  const TagToRender = as ?? getDefaultTagFromVariant(variant)

  const isMobile = useIsMobile()

  return (
    <TagToRender
      {...otherProps}
      ref={ref}
      className={clsx(
        s.Typography,
        s[`Typography_${isMobile && mobileVariant ? mobileVariant : variant}`],
        className
      )}
    >
      {children ?? text}
    </TagToRender>
  )
}

export const Typography = React.forwardRef(TypographyInner) as unknown as <
  E extends React.ElementType = 'div'
>(
  props: ITypographyProps<E> & { ref?: React.ForwardedRef<HTMLElement> }
) => ReturnType<typeof TypographyInner>
