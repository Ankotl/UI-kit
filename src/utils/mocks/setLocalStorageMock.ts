import { EnumLocalStorageKeys } from '../../enums/localStorageEnum'

export const setLocalStorageMock = () => {
  class LocalStorageMock {
    store: Record<string, string> = {}

    clear() {
      this.store = {}
    }

    setItem(key: EnumLocalStorageKeys, value: string) {
      this.store[key] = value
    }

    getItem(key: EnumLocalStorageKeys) {
      return this.store[key] || null
    }

    removeItem(key: EnumLocalStorageKeys) {
      delete this.store[key]
    }
  }

  Object.defineProperty(window, 'localStorage', {
    value: new LocalStorageMock(),
  })
}
