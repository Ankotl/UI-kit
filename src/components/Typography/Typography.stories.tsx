import React from 'react'

import { Story } from '@storybook/react'

import { getStoryMeta } from '../../utils/storybook/getStoryMeta'
import { getStoryTemplate } from '../../utils/storybook/getStoryTemplate'

import { EnumTypographyVariants, Typography } from './Typography'

const storiesMeta = getStoryMeta({
  component: Typography,
})
export default storiesMeta

/* Interactive */
const storyTemplate = getStoryTemplate(Typography)

export const Interactive = storyTemplate({
  variant: EnumTypographyVariants.H1,
  text: 'Some important message',
  as: 'h1' as React.ElementType,
})
Interactive.decorators = [
  (Story) => {
    const containerStyle = {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      minHeight: '100vh',
    }
    return (
      <div style={containerStyle}>
        <Story />
      </div>
    )
  },
]

/* ShowCase */
interface IShowCaseTemplateProps {
  variant: EnumTypographyVariants
}
const ShowCaseTemplate: React.FC<IShowCaseTemplateProps> = ({ variant }) => {
  const containerStyle = {
    boxShadow: 'rgba(100, 100, 111, 0.2) 0px 7px 29px 0px',
    padding: 32,
    borderRadius: 20,
    display: 'flex',
    alignItems: 'center',
  }

  return (
    <div style={containerStyle}>
      <Typography variant={variant} text={`Typography variant - ${variant}`} />
    </div>
  )
}

export const ShowCase: Story = () => {
  const containerStyle = {
    display: 'grid',
    gridTemplateColumns: 'repeat(2, minmax(360px, 600px))',
    gap: 18,
  }

  return (
    <div style={containerStyle}>
      {Object.values(EnumTypographyVariants)
        .sort((a, b) => (a > b ? 0 : 1))
        .map((value, i) => (
          <ShowCaseTemplate key={i} variant={value} />
        ))}
    </div>
  )
}

ShowCase.parameters = {
  controls: { hideNoControlsWarning: true },
}
