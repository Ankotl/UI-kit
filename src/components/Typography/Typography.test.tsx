import React from 'react'

import { render, screen } from '@testing-library/react'

import '../../utils/mocks/setWindowMatchMediaMock'

import {
  EnumTypographyVariants,
  getDefaultTagFromVariant,
  Typography,
} from './Typography'
import { Interactive } from './Typography.stories'

describe('Typography component test', () => {
  test('Has right states', () => {
    Object.values(EnumTypographyVariants).forEach((variant) => {
      const { container } = render(
        <Typography text={'testText'} variant={variant} />
      )
      expect(container).toMatchSnapshot()
    })

    const { container: containerSecondary } = render(
      <Typography
        text={'testDivText'}
        variant={EnumTypographyVariants.H1}
        as="div"
      />
    )
    expect(containerSecondary).toMatchSnapshot()
  })

  test('Renders content', () => {
    const testText = 'testText'
    const testTextSecond = 'testTextSecond'

    render(<Interactive {...Interactive.args} text={testText} />)
    render(<Interactive>{testTextSecond}</Interactive>)

    expect(screen.getByText(testText)).toBeTruthy()
    expect(screen.getByText(testTextSecond)).toBeTruthy()
  })

  test('Renders right html tag', () => {
    const defaultTags = Object.values(EnumTypographyVariants).map((value) =>
      getDefaultTagFromVariant(value)
    )
    const defaultTagsExpectedResult = [
      'h1',
      'h2',
      'h3',
      'h4',
      'p',
      'p',
      'p',
      'div',
    ]
    expect(defaultTags).toEqual(defaultTagsExpectedResult)

    defaultTags.forEach((tag) => {
      const { container } = render(
        <Interactive
          {...Interactive.args}
          as={tag}
          variant={EnumTypographyVariants.CAPTION}
        />
      )

      expect(container.querySelector(tag as unknown as string)).toBeTruthy()
    })
  })
})
