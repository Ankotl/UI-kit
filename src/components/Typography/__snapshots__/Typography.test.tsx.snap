// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography component test Has right states 1`] = `
<div>
  <h1
    class="Typography Typography_h1"
  >
    testText
  </h1>
</div>
`;

exports[`Typography component test Has right states 2`] = `
<div>
  <h2
    class="Typography Typography_h2"
  >
    testText
  </h2>
</div>
`;

exports[`Typography component test Has right states 3`] = `
<div>
  <h3
    class="Typography Typography_h3"
  >
    testText
  </h3>
</div>
`;

exports[`Typography component test Has right states 4`] = `
<div>
  <h4
    class="Typography Typography_h4"
  >
    testText
  </h4>
</div>
`;

exports[`Typography component test Has right states 5`] = `
<div>
  <p
    class="Typography Typography_p1"
  >
    testText
  </p>
</div>
`;

exports[`Typography component test Has right states 6`] = `
<div>
  <p
    class="Typography Typography_p2"
  >
    testText
  </p>
</div>
`;

exports[`Typography component test Has right states 7`] = `
<div>
  <p
    class="Typography Typography_p3"
  >
    testText
  </p>
</div>
`;

exports[`Typography component test Has right states 8`] = `
<div>
  <div
    class="Typography Typography_caption"
  >
    testText
  </div>
</div>
`;

exports[`Typography component test Has right states 9`] = `
<div>
  <div
    class="Typography Typography_h1"
  >
    testDivText
  </div>
</div>
`;
